### ISSUE TITLE
Add constants to the scripting language

### ISSUE THREAD

**Issue by lkcampbell** [chrismaltby/gb-studio#1553]
**Is your feature request related to a problem? Please describe.**
I use what would be considered enumerations in other languages, a lot in my code and, instead of using integers, I use global variables like pseudo-constants, to give the all those different integers human readable names.

For example:

![Screenshot 2024-09-02 at 4 25 35 AM](https://github.com/user-attachments/assets/542920f9-510e-479b-ad1f-8a4f51c9f06b)

Of course, this uses up a lot of global variables, which I would rather not do. Also, I don't have the added security of having true constants that cannot have their values changed.

**Describe the solution you'd like**
Instead of using global variables like constants, if there could be real constants that don't use up any global variable space. Internally, the constants are still represented by numbers but in the app they all use constant names so it is easy for humans to read them. It seems like a really useful feature to have in a scripting environment where so much of the data is represented by integers with a variety of purposes that are difficult to keep track of and remember.

**Describe alternatives you've considered**
I don't know enough about how the internals work to know if this is even possible, but I was thinking of the Joypad directions as an analogy of something represented as "up", "down", "left" and "right" in the app, even though they are actually integers (flags?) under the hood. Could it be done as a pre-compile process? Could it just be a human-readable view of certain number in the data model?

**Comment by maxoakland** [comment]
My thoughts on this are it would be a useful feature *but* I'm not sure how it would work on the hardware

For one thing, I think there's a limit to global variables specifically because the Gameboy has a limited memory space. So adding constants would really mean adding extra global variables that GBStudio treats as constants

I guess the key question is: How much of the Gameboy's memory space is used up? How many more variables could be stored and used as constants? Or is there another way? Could constants be stored in the ROM and read from there instead of a variable?

**Comment by lkcampbell** [comment]
I agree if constants still use up global variable space, there is no point in doing this. I will just continue to use global variables as constants. But if there is a way to have constants without using up global variable space, by substituting actual numbers anywhere the constants are used, this would be worth doing. But, yeah, I lack the technical knowledge to ascertain whether this is even possible.

**Comment by chrismaltby** [comment]
This is a good idea :-) I'm experimenting with a branch that starts to add constants support

There would be an extra panel in the navigator sidebar for managing constants
<img width="445" alt="Screenshot 2024-09-06 at 12 48 48" src="https://github.com/user-attachments/assets/960171a3-3471-47cb-8ae8-d7c74329a8fd">

Clicking on the constant lets you rename + set the value
<img width="540" alt="Screenshot 2024-09-06 at 12 59 40" src="https://github.com/user-attachments/assets/e10c4aa8-5dc0-4f80-957b-c25013a0797a">

You can use constants anywhere that either script values or expressions are supported.

Script values now have "Constants" available to select in dropdown menu
<img width="502" alt="Screenshot 2024-09-06 at 12 49 34" src="https://github.com/user-attachments/assets/bf137c14-af9a-4376-91b9-e35dd6b518d2">

Expressions let you autocomplete constants by just starting to type their names
<img width="506" alt="Screenshot 2024-09-06 at 13 03 41" src="https://github.com/user-attachments/assets/f9849971-4564-4a7c-be69-36ebe879b536">

At compile tile these are also constants
`game_globals.i`
will contain
```asm
CONST_MAX_HP = 999
CONST_ENEMY_GHOST_HP = 100
CONST_BASE_DAMAGE = 8
CONST_ENEMY_GOBLIN_HP = 60
CONST_ENEMY_ROCK_HP = 999
CONST_SWORD_DAMAGE = 30
```

and uses of constants get turned into the following GBVM
```
        VM_RPN
            .R_REF      VAR_S4A4_HP
            .R_INT16    CONST_BASE_DAMAGE
            .R_INT16    CONST_SWORD_DAMAGE
            .R_OPERATOR .RND
            .R_OPERATOR .ADD
            .R_OPERATOR .SUB
            .R_REF_SET  VAR_S4A4_HP
            .R_STOP
```

which takes uses no additional global variable space and is identical to just number values manually.

Do you think something like this would work for you?

You can test out the branch here:
https://github.com/chrismaltby/gb-studio/actions/runs/10737886132

**Comment by lkcampbell** [comment]
Yes, this is *exactly* what I was imagining and describing. And you did it so quickly.

I will take a look at it this weekend, thanks!

**Comment by lkcampbell** [comment]
I took it for a quick spin. Every thing works how I imagined it would.

My only comment, which I think might fall into separate feature request, is to allow the "When: " branches of the Switch event to check for constants. Because the Switch event is a common scenario for using pseudo-constants for me. Here's an example with this new build where I am simulating this idea with a comment event:

![Screenshot 2024-09-07 at 5 27 14 AM](https://github.com/user-attachments/assets/e05aeb4a-9b33-44a7-86aa-0bdfd2028e45)

Again, I understand that this is a separate feature request, but I wanted to get it out there while I am testing stuff so I don't forget. Let em know if you want me to submit a separate feature request for this.

Other than that, it's working great! And on a related note, the new project structure you introduced really makes it quick and easy to add Constants. Great stuff!

**Comment by chrismaltby** [comment]
Still working on this but think you're right about the Switch event, I've got a build now that lets you choose either numbers or constants for the branch conditions :-)

<img width="525" alt="Screenshot 2024-09-13 at 19 44 50" src="https://github.com/user-attachments/assets/2efd2a01-21f4-46d2-9c1f-edccc6f5d3fb">

**Comment by chrismaltby** [comment]
This has been merged into the develop branch now so should be included in the next release :-)

**Comment by lkcampbell** [comment]
@chrismaltby Very cool. I will take a look at it. Do you want me to close this issue once the feature is confirmed now, or after the release? Or is that something you prefer to do?

**Comment by lkcampbell** [comment]
This feature was added in v4.2.0. Closing the request. Thanks.
